packages/client/src/links/wsLink/wsClient/reconnectManager.ts

```diff
@@ -0,0 +1,94 @@
+import { sleep } from '@trpc/server/unstable-core-do-not-import';
+import { TRPCWebSocketClosedError } from './utils';
+import type { WsConnection } from './wsConnection';
+
+/**
+ * Represents a fatal WebSocket error that prevents reconnection attempts.
+ * This error is used when retrying the connection would consistently fail due to
+ * a non-recoverable issue.
+ */
+export class TRPCWebSocketReconnectFatal extends Error {
+  constructor(opts: { message: string; cause?: unknown }) {
+    super(opts.message, {
+      cause: opts.cause,
+    });
+    this.name = 'TRPCWebSocketReconnectFatal';
+    Object.setPrototypeOf(this, TRPCWebSocketReconnectFatal.prototype);
+  }
+}
+
+/**
+ * Manages WebSocket reconnections with configurable retry logic.
+ * Handles the lifecycle of a WebSocket connection, including error handling and retries.
+ */
+export class ReconnectManager {
+  private connection: WsConnection | undefined;
+
+  constructor(
+    private readonly retryDelayMs: (attemptIndex: number) => number,
+    private readonly callbacks: {
+      onError: (error: TRPCWebSocketClosedError) => void;
+    },
+  ) {}
+
+  public attach(connection: NonNullable<typeof this.connection>) {
+    this.connection = connection;
```

Comment:
We should likely ensure that the connection is dead before overriding it

Reply:
Fixed by making the connection immutable. The connection is now only set once in the constructor, removing the possibility of attaching a new connection later.